openvdb/tools/Count.h

```diff
@@ -0,0 +1,220 @@
+//////////////////////////////////////////////////////////////////////////
+//
+// TM & (c) Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved. This software is distributed under the
+// Mozilla Public License 2.0 ( http://www.mozilla.org/MPL/2.0/ )
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @file Count.h
+///
+/// @author Dan Bailey
+///
+/// @brief Counting tools.
+
+#ifndef OPENVDB_TOOLS_COUNT_HAS_BEEN_INCLUDED
+#define OPENVDB_TOOLS_COUNT_HAS_BEEN_INCLUDED
+
+#include <openvdb/tree/NodeManager.h>
+
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+
+/// @brief Return a vector with node counts. The number of node of type
+/// NodeType is given as element NodeType::LEVEL in the return vector.
+/// Thus, the size if this vector corresponds to the height (or depth)
+/// of this tree.
+///
+/// @param tree       the tree to be counted
+/// @param threaded   enable or disable threading (threading is enabled by default)
+template <typename TreeT>
+std::vector<Index>
+nodeCount(const TreeT& tree, bool threaded = true);
+
+
+/// @brief Return the total number of active voxels.
+///
+/// @param tree       the tree to be counted
+/// @param threaded   enable or disable threading (threading is enabled by default)
+template <typename TreeT>
+Index64
+activeVoxelCount(TreeT& tree, bool threaded = true);
+
```

Comment:
I kind of wish we could not have a threaded flag here.  How is a caller supposed to know to thread or not, when the decision to thread is dependent on active voxel count?  :>  I wonder if we could auto-set threading internally on a metric we can quickly compute.

Reply:
I know what you mean, I have a love-hate relationship with this parameter. When you're tracking down some complicated deterministic bug, being able to disable threading for a whole method and all it's sub-methods to verify exactly where the problem lies is nice. The annoying part of this is that you always want it to be the last parameter, so any new parameters added to this method always means a change in API and the process of deprecating one while calling the new method and then finally removing it. :(

I'm not sure I see it's use being set based on the theoretical size of a grid as you can only guess at this stage, but where it could perhaps be useful is for example processing 1000 grids in parallel and wanting each one to be computed on a single thread.